File: leetcode/074-Search-a-2D-Matrix/Searcha2DMatrix_001.py
class Solution:
    # @param {integer[][]} matrix
    # @param {integer} target
    # @return {boolean}
    def searchMatrix(self, matrix, target):
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return False

        m, n = len(matrix), len(matrix[0])

        l, r = 0, m * n - 1
        while l <= r:
            mid = (l + r) / 2
            i, j = mid / n, mid % n
            if matrix[i][j] == target:
                return True
            elif matrix[i][j] < target:
                l += 1
            else:
                r -= 1

        return False

File: lintcode/074-028-Search-a-2D-Matrix/SearchA2dMatrix_001.java
public class Solution {
    /**
     * @param matrix, a list of lists of integers
     * @param target, an integer
     * @return a boolean, indicate whether matrix contains target
     */
    public boolean searchMatrix(int[][] matrix, int target) {
        // write your code here
        if (matrix.length == 0 || matrix[0].length == 0) {
            return false;
        }

        int nrows = matrix.length;
        int ncols = matrix[0].length;
        int l = 0;
        int r = nrows * ncols - 1;
        int m;
        int i;
        int j;

        while (l <= r) {
            m = (l + r) / 2;
            i = m / ncols;
            j = m % ncols;
            if (matrix[i][j] == target) {
                return true;
            } else if (matrix[i][j] < target) {
                l = m + 1;
            } else {
                r = m - 1;
            }
        }

        return false;
    }
}